### DISCUSSION TITLE
Support for markdown callouts

### DISCUSSION THREAD

**Discussion by Adamkadaban** [gettalong/kramdown#805] [General]
Are there any plans to support markdown callouts like these?

> [!NOTE]
> Highlights information that users should take into account, even when skimming.

> [!TIP]
> Optional information to help a user be more successful.

> [!IMPORTANT]
> Crucial information necessary for users to succeed.

> [!WARNING]
> Critical content demanding immediate user attention due to potential risks.

> [!CAUTION]
> Negative potential consequences of an action.

They're supported on github, gitbook, obsidian, and probably more. This addition would make compatibility a bit easier.

I currently have a jekyll plugin for this but it's not very well made and requires liquid templates
https://github.com/Adamkadaban/jekyll-alerts-plugin

**Comment by gettalong** [comment]
As far as I can see the syntax is just plain Markdown but the interpretation from standard Markdown is different, i.e. the output is different.

If you want to have custom HTML output for this, you need to subclass the HTML converter and adapt the [`convert_blockquote`](https://github.com/gettalong/kramdown/blob/master/lib/kramdown/converter/html.rb#L141) method appropriately.

**Reply by benpickles** [reply]
Is this something that could live in kramdown-parser-gfm, maybe even enabled by default? It feels like perhaps not because it specifically says "parser" and not "converter" but...

**Reply by gettalong** [reply]
The GFM parser could also detect those lines and create different AST nodes to represent the callouts. Then the converter need not be changed. I just mentioned the converter because it is much each easier to adapt it to ones needs.

**Reply by benpickles** [reply]
I took your pointer and monkeypatched the converter, you can see it in action here https://parklife.dev/colophon (source: https://github.com/benpickles/parklife.dev/blob/main/lib/kramdown_enhancements.rb).

It was only once I'd implemented the change that I understood why it's not quite so simple to add support to Kramdown: because the ultimate output is entirely dependent on the wants of the developer. In my case I wanted to output an `<aside>` and a heading, and include icons and colours, whereas GitHub output a `<div>` and some `<p>`s (and icons and colours), others might want something completely different.

I haven't looked at the parser, I'm guessing it might be a better place to put things so that a specific alert callout converter method could be defined instead of overloading the `convert_blockquote` method. But I wouldn't know where to start... Any tips?!

BTW working within Kramdown for this really felt like the right place. I've previously either modified the Markdown before converting or the HTML after converting but working with Kramdown `Element`s felt a lot more healthy.

**Reply by gettalong** [reply]
In the parser you could define a new element type `:callout` and insert that instead of the blockquote into the AST. Then each converter would need to define a `convert_callout` method that converts the callout into the required format.
